File: tic-tac-toe-javafx-java/src/main/java/com/mj/tic/tac/toe/javafx/java/util/GameState.java
package com.mj.tic.tac.toe.javafx.java.util;

/**
 * Defines the states of a tic-tac-toe game session.
 * <p>
 * This enum represents the high-level lifecycle states of the game, controlling
 * transitions between the start of a new session, individual match resets,
 * session termination, and the option to replay.
 * </p>
 *
 * @author Montaser Jamal
 * @email montaser.jjs@gmail.com
 * @phone +962-786258874
 * @created 28-05-2026
 * @since 1.0.0
 */
public enum GameState {

    RESET_GAME,
    NEW_GAME,
    GAME_OVER;
}

File: tic-tac-toe-javafx-kotlin/src/main/kotlin/com/mj/tic/tac/toe/javafx/kotlin/util/GameState.kt
package com.mj.tic.tac.toe.javafx.kotlin.util

/**
 * Represents the lifecycle states of a game session.
 *
 * These states are published through [GameStateManager] to all
 * registered [Subscriber] instances. Each state carries an optional
 * typed payload delivered to subscribers via the publish method.
 *
 * @author Montaser Sbaih
 * @version 1.0
 * @email montaser.jjs@gmail.com
 * @phone +962-786258874
 * @since 29-05-2026
 */

enum class GameState {
    /**
     * Published when the game is fully reset.
     *
     * The board is cleared but score counters are preserved.
     * Payload: none (null).
     */
    RESET_GAME,

    /**
     * Published when a new game begins with a chosen difficulty.
     *
     * Payload: [Difficulty] — the difficulty level chosen by the player.
     */
    NEW_GAME,

    /**
     * Published when the game reaches a terminal state.
     *
     * Payload: [Player]? — the winning player ([Player.HUMAN] or
     * [Player.COMPUTER]), or null if the game ended in a draw.
     */
    GAME_OVER
}
